server-protocol.md

```diff
@@ -104,6 +118,50 @@ was sent in the wrong order), an `error` response will be sent, This response
 will include the error string in the `error` key, and a full copy of the
 original message dictionary in `orig`.
 
+
+## Permission to Use the Server
+
+Server operators may wish to deny service to some clients. We
+generally refer to this as "Permission" and imagine future additions
+of use-cases and methods, although only one such pair is decribed
+currently.
+
+One such use-case is if the server is under a Denial of Service (DoS)
+attack or other malicious activity. We describe a "proof of work"
+scheme for clients to gain permission to use a server under DoS
+attack.
+
+In the `welcome` message the server may include a `permission-required` key. If provided, this will point to a `dict` with one key per supported method along with any method-specific options.
+For example:
+
+    {
+        "none": {}
+        "hashcash": {
+            "bits": 6,
+            "resource": "arbitrary string"
+        }
+    }
+
+Currently, the following are supported by the protocol:
+
+* `none`: no permission required, send a normal `bind`.
+* `hashcash`: Includes keys for `bits` and `resource` used as input as per the
+[Hashcash](http://hashcash.org) specifications. The client must
+include a reply that conforms to the [Hashcash v1](http://hashcash.org/docs/hashcash.html#stamp_format__version_1_)
+format in the `submit-permissions` part of the `bind` message:
+
+    {
+        "type": "hashcash",
+        "stamp": "1:6:210623:arbitrary string::9WrCxB1SdCBOM3i5:000005"
+    }
+
+The above stamp was generated with `hashcash -b 6 -m "arbitrary string"`.
+
+More methods may be added to the protocol in future. Clients must
+ignore methods they do not support. Clients may choose any supported
+method.
+
```

Comment:
We should add a note that any bearer-token -like scheme (including `hashcash`) must be run over a secure (TLS) connection, otherwise a network attacker can use your hard-earned stamp for themselves.

The public relay server is currently using Caddy to accept StartTLS, so clients can use `wss:` instead of `ws:`, but I don't think I've updated the client code's default URL yet.

Reply:
I think this is a good reminder that we actually want to phase out un-encrypted communication, even on the transport layer and if the relevant content is already encrypted.

Reply:
What exactly is our threat model regarding the client-server connection in general? Because a token can be used only once, an attacker would either need a really low-latency connection or be a man in the middle. 

However, this is a lot more critical when using proper authentication methods instead of just proof of work. We should add that server implementations must not allow any authentication methods that are subject to replay attacks over an un-encrypted connection.

Reply:
For non-permission things, the client-server connection should be protected mainly to improve user privacy. If they're unencrypted, then a network attacker (say, the person controlling your coffee-shop wifi) has the same power as the mailbox server: they can tell which nameplate you're using, so they can definitively correlate your connection with somebody else's (if they're watching that network too), or then can attempt to guess your code (once).

If that connection is encrypted, they can only infer the correlation, by looking at timing (granted, this is a pretty strong inference, because our protocol messages are pretty lock-step). And they can only infer the nameplate you're using (by querying the mailbox server with a `list` command) frequently and watching for the new one to appear. (again, this is pretty easy).

I've been lazy about moving connections over to TLS because the difference in attacker power is so small. But it's the right thing to do, in general.

For permission things, yeah, the threat is either a race (the attacker waiting for the user to do the work and then consuming their token first), or a full MitM where the user always loses the race because the attacker just disconnects them.

Reply:
If we do a push towards TLS anyways, how about specifying that servers that require usage permission (other than `none`) must do this over a secure connection? If the connection is not secure but the server wants a permission, it errors out.